xds/internal/balancer/ringhash/picker_test.go

```diff
@@ -142,7 +146,154 @@ func (s) TestPickerPickFirstTwo(t *testing.T) {
+				t.Errorf("Pick() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got.SubConn != tt.wantSC {
+				t.Errorf("Pick() got = %v, want picked SubConn: %v", got, tt.wantSC)
+			}
+			if sc := tt.wantSCToConnect; sc != nil {
+				select {
+				case <-sc.(*testutils.TestSubConn).ConnectCh:
+				case <-time.After(defaultTestShortTimeout):
+					t.Errorf("timeout waiting for Connect() from SubConn %v", sc)
+				}
+			}
+		})
+	}
+}
+
+func (s) TestPickerNoRequestHash(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	ring, epStates := testRingAndEndpointStates([]connectivity.State{connectivity.Ready})
+	p := &picker{
+		ring:           ring,
+		logger:         internalgrpclog.NewPrefixLogger(logger, "test-ringhash-picker"),
+		endpointStates: epStates,
+	}
+	_, err := p.Pick(balancer.PickInfo{
+		Ctx: ctx,
+	})
+	if err == nil {
+		t.Errorf("Pick() should have failed with no request hash")
+	}
+}
+
+func (s) TestPickerRequestHashKey(t *testing.T) {
+	tests := []struct {
+		name         string
+		headerValues []string
+		expectedPick int
+	}{
+		{
+			name:         "header not set",
+			expectedPick: 0, // Random hash set to 0, which is within ]MaxUint64 / 3 * 2, 0]
+		},
+		{
+			name:         "header empty",
+			headerValues: []string{""},
+			expectedPick: 0, // xxhash.Sum64String("value1,value2") is within ]MaxUint64 / 3 * 2, 0]
+		},
+		{
+			name:         "header set to one value",
+			headerValues: []string{"some-value"},
+			expectedPick: 1, // xxhash.Sum64String("some-value") is within ]0, (MaxUint64 / 3]
+		},
+		{
+			name:         "header set to multiple values",
+			headerValues: []string{"value1", "value2"},
+			expectedPick: 2, // xxhash.Sum64String("value1,value2") is within ](MaxUint64 / 3), (MaxUint64 / 3 * 2)]
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+
+			ring, epStates := testRingAndEndpointStates(
+				[]connectivity.State{
+					connectivity.Ready,
+					connectivity.Ready,
+					connectivity.Ready,
+				})
+			headerName := "some-header"
+			p := &picker{
+				ring:              ring,
+				logger:            internalgrpclog.NewPrefixLogger(logger, "test-ringhash-picker"),
+				endpointStates:    epStates,
+				requestHashHeader: headerName,
+				randUint64:        func() uint64 { return 0 },
+			}
+			for _, v := range tt.headerValues {
+				ctx = metadata.AppendToOutgoingContext(ctx, headerName, v)
+			}
+			res, err := p.Pick(balancer.PickInfo{
+				Ctx: ctx,
+			})
+			if err != nil {
+				t.Errorf("Pick() failed: %v", err)
+			}
+			if res.SubConn != testSubConns[tt.expectedPick] {
+				t.Errorf("Pick() got = %v, want SubConn: %v", res.SubConn, testSubConns[tt.expectedPick])
+			}
+		})
+	}
+}
+
+func (s) TestPickerRandomHash(t *testing.T) {
```

Comment:
Can any of these tests be written in e2e style? In the past I've seen that some of these unit level tests on the LB policies fail to achieve the end goal and end up leaving bugs in place. I understand that writing e2e style tests would be more work, but if that is possible, even if it can captured in an issue with the exact scenarios, we can get someone else to work on it.

Reply:
I agree with your sentiment that end to end tests would be nicer for this, however the relyiance on a random value to pick a backend makes them a bit hard to write. In fact I tried to write some, but didn't manage to write any really valuable test case, since it's not possible to know in advance how many endpoints will end up being woken up on each RPC (although it should be capped to 2 -- maybe that'd be a valuable test actually, thinking more about it). Let me try to implement that.

Reply:
Thanks.